Paint worklets: Implement timeout for worklet painter threads

yati-sagade opened the pull request:
When a paint worklet thread takes too long, we would like to move on,
since we have a ~16ms budget for rendering at 60fps. At the moment, there
is no provision in the paintworklet spec to signal such timeouts to the
developer. ajeffrey opened an [issue][1] for this, but it got punted to
v2 of the spec. Hence we are silently timing out unresponsive paint
scripts.

The timeout value is chosen to be 10ms by default, and can be overridden
by setting the SERVO_PAINT_WORKLET_TIMEOUT_MS environment variable.

In the absence of such a timeout, the reftest in this commit would fail
by timing out the testrunner itself, since the paint script never
returns. From my discussions with ajeffrey, this should do until we spec
out a way to signal timeouts to the script developer.

This fixes #17370.

[1]: https://github.com/w3c/css-houdini-drafts/issues/507

<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/19256)
<!-- Reviewable:end -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @metajack (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/paintworkletglobalscope.rs, components/script_traits/lib.rs, components/script_traits/lib.rs
 * @fitzgen: components/script/dom/paintworkletglobalscope.rs, components/script_traits/lib.rs, components/script_traits/lib.rs
 * @emilio: components/layout/display_list_builder.rs

yati-sagade committed:
Paint worklets: Implement timeout for worklet painter threads

When a paint worklet thread takes too long, we would like to move on,
since we have a ~16ms budget for rendering @60fps. At the moment, there
is no provision in the paintworklet spec to signal such timeouts to the
developer. ajeffrey opened an [issue][1] for this, but it got punted to
v2 of the spec. Hence we are silently timing out unresponsive paint
scripts.

The timeout value is chosen to be 10ms by default, and can be overridden
by setting the `dom.worklet.timeout_ms` pref.

In the absence of such a timeout, the reftest in this commit would fail
by timing out the testrunner itself, since the paint script never
returns. From my discussions with ajeffrey, this should do until we spec
out a way to signal timeouts to the script developer.

Since we did not have a better way to trigger a timeout than a busy
waiting loop (which would hog one core of the test machine until the
timeout was reached), we decided to implement a test only blocking
sleep, available to the PaintWorkletGlobalScope. Since
`dom.worklet.enabled` enables worklets in general, we also decided to
have another pref `dom.worklet.blockingsleep.enabled`, which, in
addition to `dom.worklet.enabled`, would be required for the blocking
sleep to be available.

This fixes #17370.

[1]: https://github.com/w3c/css-houdini-drafts/issues/507

yati-sagade commented:
@metajack sorry for the delay in fixing tidy complaints :) Don't know if new builds notify you; Can you have a look please?

jdm commented:
r? @asajeffrey

asajeffrey commented on tests/wpt/mozilla/tests/mozilla/worklets/test_paint_worklet_timeout.js:
```diff
@@ -0,0 +1,5 @@
+registerPaint("testgreen", class {
+    paint(ctx, size) {
+        for(;;); // take a bit too long to do initialization
```
It would be nice if there was a blocking API we could use here rather than busy-waiting.

yati-sagade replied:
Hey, I tried playing with a _synchronous_ XHR to `/`, but that doesn't seem to trigger the timeout (the request still returns in time for the timeout). Do you have any other suggestions for such an API? I've pushed changes to cover the rest of your suggestions!

asajeffrey replied:
I think the XHR API isn't available to worklets unfortunately. Can you check to see if you're getting an error from XHR being undefined or if the HTTP request is going through. If it's the latter, there may be some wpt cleverness for getting the HTTP server to delay responding.

jdm replied:
XHR should not even be defined in worklet globals; if it is, that's an issue that should be filed :)

yati-sagade replied:
@jdm ack, indeed it seems it isn't available. For some reason I am not seeing error messages when I run the debug profile (but can see log statements). However, my paint() function doesn't execute any subsequent log statements after I use `window.XMLHttpRequest` or `XMLHttpRequest`.

Any other ideas for a slow/blocking API?

asajeffrey replied:
If we had the `Atomics` API then we could use `wait`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics/wait but I don't think we do :/ I think we're stuck with a busy-wait. @jdm, should we check in the busy-wait, even though it will probably thrash one core of the test machine until the wpt test times out?

jdm replied:
Is there a test-only API we could implement that would do what we want?

asajeffrey replied:
We could implement something like synchronous sleep, and hide it behind a pref.

yati-sagade replied:
That's a good idea. I can also try implementing this, if you show me the pieces of a similar existing API.

asajeffrey replied:
There's a test-only worklet API in https://github.com/servo/servo/blob/master/components/script/dom/webidls/TestWorkletGlobalScope.webidl, and a test-only iterable interface in https://github.com/servo/servo/blob/master/components/script/dom/webidls/TestBindingIterable.webidl. These should be good models to follow.

asajeffrey commented on tests/wpt/mozilla/tests/mozilla/worklets/test_paint_worklet_timeout.html:
```diff
@@ -0,0 +1,31 @@
+<!doctype html>
```
Yay tests!

asajeffrey reviewed (commented):
Looks great, just a few niggles!

asajeffrey commented:
Sorry about the delay!

asajeffrey commented on tests/wpt/mozilla/tests/mozilla/worklets/test_paint_worklet_timeout.js:
```diff
@@ -0,0 +1,7 @@
+    paint(ctx, size) {
+        ctx.fillStyle = 'green';
+        ctx.fillRect(0, 0, size.width, size.height);
+        sleep(30); // too long for a paintworklet to init
```
YAY!

asajeffrey reviewed (approved):
This all looks really good, thanks for all the effort you put in!

asajeffrey commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 1146bfd has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey 1146bfd23509e527966f8fab422b0b7c5841b0a6 -->

bors-servo commented:
:hourglass: Testing commit 1146bfd23509e527966f8fab422b0b7c5841b0a6 with merge d3cebd723c066800f9b66d5da686505bade3a9dd...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/10208)

tigercosmos commented:
please run `./mach update-manifest` and rebase the commit change.

yati-sagade commented:
@tigercosmos done!

KiChjang commented:
@bors-servo r=asajeffrey

bors-servo commented:
:pushpin: Commit aa48a2c has been approved by `asajeffrey`

<!-- @bors-servo r=asajeffrey aa48a2c2e3c7699a167d9ffe791f4bb17e9b9f1c -->

bors-servo commented:
:hourglass: Testing commit aa48a2c2e3c7699a167d9ffe791f4bb17e9b9f1c with merge 54f8cc37c5a7874607128e3518fd0982f7065565...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/10209), [arm32](http://build.servo.org/builders/arm32/builds/10445), [arm64](http://build.servo.org/builders/arm64/builds/10473), [linux-dev](http://build.servo.org/builders/linux-dev/builds/10499), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/7340), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/7364), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/10565), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/2987), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/3002), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/7251), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/7311), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/3069), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/3024), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/5866)
Approved by: asajeffrey
Pushing 54f8cc37c5a7874607128e3518fd0982f7065565 to master...

yati-sagade commented:
\o/ thanks @asajeffrey for all the help
